File: 4.Authentication and authorization for backend engineers/code/typescript/05_json_web_tokens_jwt_3.ts
import jwt from 'jsonwebtoken';

const secret = 'keep-this-very-secret';

// mint a self-contained token carrying the claims
export function sign(userID: string, role: string): string {
    const claims = {
        sub: userID,                               // user id
        role: role,                                // for authorization
        iat: Math.floor(Date.now() / 1000),        // issued at
        exp: Math.floor(Date.now() / 1000) + 3600  // expiry (1 hour)
    };
    return jwt.sign(claims, secret, { algorithm: 'HS256' });
}

// verify signature + expiry; returns the claims if valid
export function verify(tokenStr: string): any {
    try {
        // algorithms restricts to HS256 to stop "alg: none" attacks
        return jwt.verify(tokenStr, secret, { algorithms: ['HS256'] });
    } catch (err) {
        throw new Error('Invalid signature or expired token');
    }
}

File: 4.Authentication and authorization for backend engineers/code/rust/05_json_web_tokens_jwt_3.rs
use jsonwebtoken::{encode, decode, Header, Validation, EncodingKey, DecodingKey, errors::Error};
use serde::{Deserialize, Serialize};
use std::time::{SystemTime, UNIX_EPOCH};

const SECRET: &[u8] = b"keep-this-very-secret";

#[derive(Debug, Serialize, Deserialize)]
pub struct Claims {
    pub sub: String, // user id
    pub role: String, // for authorization
    pub iat: u64, // issued at
    pub exp: u64, // expiry
}

// mint a self-contained token carrying the claims
pub fn sign(user_id: &str, role: &str) -> Result<String, Error> {
    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
    let claims = Claims {
        sub: user_id.to_string(),
        role: role.to_string(),
        iat: now,
        exp: now + 3600,
    };
    
    encode(&Header::default(), &claims, &EncodingKey::from_secret(SECRET))
}

// verify signature + expiry; returns the claims if valid
pub fn verify(token_str: &str) -> Result<Claims, Error> {
    // pin the algorithm to stop "alg: none" attacks
    let mut validation = Validation::new(jsonwebtoken::Algorithm::HS256);
    validation.validate_exp = true;
    
    let token_data = decode::<Claims>(
        token_str,
        &DecodingKey::from_secret(SECRET),
        &validation,
    )?;
    
    Ok(token_data.claims)
}
